service/efsPort.py

```diff
@@ -0,0 +1,630 @@
+                    "energyNeutralizerAmount"
+                ], stats, mod)
+            elif mod.item.groupID in [modGroupData["Remote Shield Booster"]["id"],
+                                      modGroupData["Ancillary Remote Shield Booster"]["id"]]:
+                stats["type"] = "Remote Shield Booster"
+                EfsPort.attrDirectMap(["shieldBonus"], stats, mod)
+            elif mod.item.groupID in [modGroupData["Remote Armor Repairer"]["id"],
+                                      modGroupData["Ancillary Remote Armor Repairer"]["id"]]:
+                stats["type"] = "Remote Armor Repairer"
+                EfsPort.attrDirectMap(["armorDamageAmount"], stats, mod)
+            elif mod.item.groupID == modGroupData["Warp Scrambler"]["id"]:
+                stats["type"] = "Warp Scrambler"
+                EfsPort.attrDirectMap(["activationBlockedStrenght", "warpScrambleStrength"], stats, mod)
+            elif mod.item.groupID == modGroupData["Target Painter"]["id"]:
+                stats["type"] = "Target Painter"
+                EfsPort.attrDirectMap(["signatureRadiusBonus"], stats, mod)
+            elif mod.item.groupID == modGroupData["Sensor Dampener"]["id"]:
+                stats["type"] = "Sensor Dampener"
+                EfsPort.attrDirectMap(["maxTargetRangeBonus", "scanResolutionBonus"], stats, mod)
+            elif mod.item.groupID == modGroupData["ECM"]["id"]:
+                stats["type"] = "ECM"
+                EfsPort.attrDirectMap([
+                    "scanGravimetricStrengthBonus", "scanMagnetometricStrengthBonus",
+                    "scanRadarStrengthBonus", "scanLadarStrengthBonus",
+                ], stats, mod)
+            elif mod.item.groupID == modGroupData["Burst Jammer"]["id"]:
+                stats["type"] = "Burst Jammer"
+                mod.itemModifiedAttributes["maxRange"] = mod.itemModifiedAttributes["ecmBurstRange"]
+                EfsPort.attrDirectMap([
+                    "scanGravimetricStrengthBonus", "scanMagnetometricStrengthBonus",
+                    "scanRadarStrengthBonus", "scanLadarStrengthBonus",
+                ], stats, mod)
+            elif mod.item.groupID == modGroupData["Micro Jump Drive"]["id"]:
+                stats["type"] = "Micro Jump Drive"
+                mod.itemModifiedAttributes["maxRange"] = 0
+                EfsPort.attrDirectMap(["moduleReactivationDelay"], stats, mod)
+            if mod.itemModifiedAttributes["maxRange"] is None:
+                pyfalog.error("Projected module {0} has no maxRange".format(mod.item.name))
+            stats["optimal"] = mod.itemModifiedAttributes["maxRange"] or 0
+            stats["falloff"] = mod.itemModifiedAttributes["falloffEffectiveness"] or 0
+            EfsPort.attrDirectMap(["duration", "capacitorNeed"], stats, mod)
+            projections.append(stats)
+        return projections
+
+    @staticmethod
+    def getModuleNames(fit):
```

Comment:
Think this was mentioned in the related issue, might be better to use a list of objects that represent the module, rather than a list of names. So, something like

	[{
		typeID: 1234,
		typeName: "Module Name",
		slot: 'high',
		position: 3,
		...	
	},
	...
	]

That would avoid having blank lines,m or lines labeling the rack. The data above could be used to reconstruct the same order, with knowledge of what rack it's in, and would be more universally applicable for other's who might look to utilize this format.

Reply:
I'm not sure if you'd read my response to this in #1624 when you wrote this?
Due to performance constraints this is almost laughably bad from an EFS standpoint.
For context I'll mention that EFS runs in the browser.

I would much prefer to make a secondary format with another name if need be.

Reply:
I did not read the comment before (they didn't show up on my GitHub notifications???), however still a bit confused.

Obviously I don't know how EFS is handling that stuff at all, so this is all based on assumptions :P As I understand it, you're using the current mod names array as a simple reminder for the user. My confusion is the assumption that you can still build that same thing fairly easily if the export was in an array-of-objects format, regardless of if it's browser-based or not - just loop though the array, ordered by slot and position, and then build the mod name array from it. It indeed would increase the size of the export, but that isn't too much of a concern (at least for pyfa and I would assume the end user, unsure if there's some limitations on EFS's side)

I'm genuinely curious as to why doing it that way would carry huge performance impacts on EFS? You would know better than I, so of course I defer to you as to what's best in this case. I simply feel that having a list of objects would make the format more widely accessible -- but you might be right; I might have my head wrapped around the idea that this format should carry all this info, and that's probably just not it's intended purpose 😛

Reply:
Overall the typical use case for EFS involves far more fits than pyfa. To reduce the barrier to usefulness, make it viable on mobile and minimize tedium I intend for EFS to have several default fits for each ship.

Eve has a lot of ships so even a very modest number per ship makes for a fairly high total. For most situations increasing the size of the format by 50% has a very similar performance impact on EFS to adding 50% more fits. If those are my options I'll pick more default fits unless the number is already really high. If the change is smaller or it enhances the simulation it's not really a big deal but this is 50-60% and mostly formatting fluff.

In terms of design you more or less hit the nail on the head. It's just not it's intended purpose. Ultimately the main purpose of this object is to construct virtual ships. Thousands of these virtual ships have to be able to fight each other at up to a thousand times normal speed. There's no point in the exports having too much detail when the simulation can't ever use it without slowing to a crawl. I intentionally abstracted out the concept of modules in the first place for that exact reason.

Hopefully this clears things up. I can go into a more technical explanation if you want but I feel like most of it will be telling you things you know alongside long winded justifications for implementation choices that wouldn't change my overall opinion even if they weren't the case.